fasd.1
.TH "FASD" "1" "Jul 16, 2012" "fasd user manual" ""
.SH NAME
.PP
fasd \- quick access to files and directories
.SH SYNOPSIS
.PP
fasd [options] [query ...]
.PP
[f|a|s|d|z] [options] [query ...]
.PP
fasd [\-A|\-D] [paths ...]
.SH OPTIONS
.IP
.nf
\f[C]
\-s\ \ \ \ \ \ \ \ \ list\ paths\ with\ ranks
\-l\ \ \ \ \ \ \ \ \ list\ paths\ without\ ranks
\-i\ \ \ \ \ \ \ \ \ interactive\ mode
\-e\ <cmd>\ \ \ set\ command\ to\ execute\ on\ the\ result\ file
\-b\ <name>\ \ only\ use\ <name>\ backend
\-B\ <name>\ \ add\ additional\ backend\ <name>
\-a\ \ \ \ \ \ \ \ \ match\ files\ and\ directories
\-d\ \ \ \ \ \ \ \ \ match\ directories\ only
\-f\ \ \ \ \ \ \ \ \ match\ files\ only
\-r\ \ \ \ \ \ \ \ \ match\ by\ rank\ only
\-t\ \ \ \ \ \ \ \ \ match\ by\ recent\ access\ only
\-R\ \ \ \ \ \ \ \ \ reverse\ listing\ order
\-h\ \ \ \ \ \ \ \ \ show\ a\ brief\ help\ message
\-[0\-9]\ \ \ \ \ select\ the\ nth\ entry
\f[]
.fi
.SH DESCRIPTION
.PP
Fasd keeps track of files and directories you access in your shell and
gives you quick access to them.
You can use fasd to reference files or directories by just a few key
identifying characters.
You can use fasd to boost your command line productivity by defining
your own aliases to launch programs on files or directories.
Fasd, by default, provides some basic aliases, including a shell
function "z" that resembles the functionality of "z" and "autojump."
.PP
The name "fasd" comes from the default suggested aliases
\f[C]f\f[](files), \f[C]a\f[](files/directories),
\f[C]s\f[](show/search/select), \f[C]d\f[](directories).
.PP
Fasd ranks files and directories by "frecency," that is, by both
"frequency" and "recency." The term "frecency" was first coined by
Mozilla and used in Firefox.
.SH EXAMPLES
.IP
.nf
\f[C]
z\ bundle
f\ \-e\ vim\ nginx\ conf
f\ \-i\ rc$
vi\ `f\ nginx\ conf`
cp\ update.html\ `d\ www`
open\ `sf\ pdf`
\f[]
.fi
.SH SHELL INITIALIZATION
.PP
To get fasd working in a shell, some initialization code must be run.
Put lines below in your POSIX compatible shell rc.
.IP
.nf
\f[C]
eval\ "$(fasd\ \-\-init\ auto)"
\f[]
.fi
.PP
This will setup a command hook that executes on every command and
advanced tab completion for zsh and bash.
.PP
If you want more control over what gets into your shell environment, you
can pass customized set of arguments to \f[C]fasd\ \-\-init\f[].
.IP
.nf
\f[C]
zsh\-hook\ \ \ \ \ \ \ \ \ \ \ \ \ #\ define\ _fasd_preexec\ and\ add\ it\ to\ zsh\ preexec\ array
zsh\-ccomp\ \ \ \ \ \ \ \ \ \ \ \ #\ zsh\ command\ mode\ completion\ definitions
zsh\-ccomp\-install\ \ \ \ #\ setup\ command\ mode\ completion\ for\ zsh
zsh\-wcomp\ \ \ \ \ \ \ \ \ \ \ \ #\ zsh\ word\ mode\ completion\ definitions
zsh\-wcomp\-install\ \ \ \ #\ setup\ word\ mode\ completion\ for\ zsh
bash\-hook\ \ \ \ \ \ \ \ \ \ \ \ #\ add\ hook\ code\ to\ bash\ $PROMPT_COMMAND
bash\-ccomp\ \ \ \ \ \ \ \ \ \ \ #\ bash\ command\ mode\ completion\ definitions
bash\-ccomp\-install\ \ \ #\ setup\ command\ mode\ completion\ for\ bash
posix\-alias\ \ \ \ \ \ \ \ \ \ #\ define\ aliases\ that\ applies\ to\ all\ posix\ shells
posix\-hook\ \ \ \ \ \ \ \ \ \ \ #\ setup\ $PS1\ hook\ for\ shells\ that\[aq]s\ posix\ compatible
tcsh\-alias\ \ \ \ \ \ \ \ \ \ \ #\ define\ aliases\ for\ tcsh
tcsh\-hook\ \ \ \ \ \ \ \ \ \ \ \ #\ setup\ tcsh\ precmd\ alias
\f[]
.fi
.PP
Example for a minimal zsh setup (no tab completion):
.IP
.nf
\f[C]
eval\ "$(fasd\ \-\-init\ posix\-alias\ zsh\-hook)"
\f[]
.fi
.PP
Note that this method will slightly increase your shell start\-up time,
since calling binaries has overhead.
You can cache fasd init code if you want minimal overhead.
Example code for bash (to be put into .bashrc):
.IP
.nf
\f[C]
fasd_cache="$HOME/.fasd\-init\-bash"
if\ [\ "$(command\ \-v\ fasd)"\ \-nt\ "$fasd_cache"\ \-o\ !\ \-s\ "$fasd_cache"\ ];\ then
\ \ fasd\ \-\-init\ posix\-alias\ bash\-hook\ bash\-ccomp\ bash\-ccomp\-install\ >|\ "$fasd_cache"
fi
source\ "$fasd_cache"
unset\ fasd_cache
\f[]
.fi
.PP
Optionally, if you can also source \f[C]fasd\f[] if you want
\f[C]fasd\f[] to be a shell function instead of an executable.
.PP
You can tweak initialization code.
For instance, if you want to use "c" instead of "z" to do directory
jumping, you can use the alias below:
.IP
.nf
\f[C]
alias\ c=\[aq]fasd_cd\ \-d\[aq]
#\ `\-d\[aq]\ option\ present\ for\ bash\ completion
#\ function\ fasd_cd\ is\ defined\ in\ posix\-alias
\f[]
.fi
.SH MATCHING
.PP
Fasd has three matching modes: default, case\-insensitive, and fuzzy.
.PP
For a given set of queries (the set of command\-line arguments passed to
fasd), a path is a match if and only if:
.IP "1." 3
Queries match the path in order.
.IP "2." 3
The last query matches the last segment of the path.
.PP
If no match is found, fasd will try the same process ignoring case.
If still no match is found, fasd will allow extra characters to be
placed between query characters for fuzzy matching.
.PP
Tips:
.IP \[bu] 2
If you want your last query not to match the last segment of the path,
append `/\[aq] as the last query.
.IP \[bu] 2
If you want your last query to match the end of the filename, append
`$\[aq] to the last query.
.SH COMPATIBILITY
.PP
Fasd\[aq]s basic functionalities are POSIX compliant, meaning that you
should be able to use fasd in all POSIX compliant shells.
Your shell need to support command substitution in $PS1 in order for
fasd to automatically track your commands and files.
This feature is not specified by the POSIX standard, but it\[aq]s
nonetheless present in many POSIX compliant shells.
In shells without prompt command or prompt command substitution (tcsh
for instance), you can add entries manually with "fasd \-A".
You are very welcomed to contribute shell initialization code for not
yet supported shells.
.SH TAB COMPLETION
.PP
Fasd offers two completion modes, command mode completion and word mode
completion.
Command mode completion works in bash and zsh.
Word mode completion only works in zsh.
.PP
Command mode completion is just like completion for any other commands.
It is triggered when you hit tab on a fasd command or its aliases.
Under this mode your queries can be separated by a space.
Tip: if you find that the completion result overwrites your queries,
type an extra space before you hit tab.
.PP
Word mode completion can be triggered on \f[I]any\f[] command.
Word completion is triggered by any command line argument that starts
with "," (all), "f," (files), or "d," (directories), or that ends with
",," (all), ",,f" (files), or ",,d" (directories).
Examples:
.IP
.nf
\f[C]
$\ vim\ ,rc,lo<Tab>
$\ vim\ /etc/rc.local

$\ mv\ index.html\ d,www<Tab>
$\ mv\ index.html\ /var/www/
\f[]
.fi
.PP
There are also three zle widgets: "fasd\-complete", "fasd\-complete\-f",
"fasd\-complete\-d".
You can bind them to keybindings you like:
.IP
.nf
\f[C]
bindkey\ \[aq]^X^A\[aq]\ fasd\-complete\ \ \ \ #\ C\-x\ C\-a\ to\ do\ fasd\-complete\ (files\ and\ directories)
bindkey\ \[aq]^X^F\[aq]\ fasd\-complete\-f\ \ #\ C\-x\ C\-f\ to\ do\ fasd\-complete\-f\ (only\ files)
bindkey\ \[aq]^X^D\[aq]\ fasd\-complete\-d\ \ #\ C\-x\ C\-d\ to\ do\ fasd\-complete\-d\ (only\ directories)
\f[]
.fi
.SH BACKENDS
.PP
Fasd can take advantage of different sources of recent / frequent files.
Most desktop environments (such as OS X and Gtk) and some editors (such
as Vim) keep a list of accessed files.
Fasd can use them as additional backends if the data can be converted
into fasd\[aq]s native format.
Below is a list of available backends.
.IP \[bu] 2
spotlight: OSX spotlight, provides entries that are changed today or
opened within the past month
.IP \[bu] 2
recently\-used: GTK\[aq]s recently\-used file (Usually available on
Linux)
.IP \[bu] 2
current: Provides everything in $PWD (whereever you are executing
\f[C]fasd\f[])
.IP \[bu] 2
viminfo: Vim\[aq]s editing history, useful if you want to define an
alias just for editing things in vim
.PP
You can define your own backend by declaring a function by that name in
your \f[C]\&.fasdrc\f[].
You can set default backend with \f[C]_FASD_BACKENDS\f[] variable in our
\f[C]\&.fasdrc\f[].
.SH TWEAKS
.PP
Upon every execution, fasd will source "/etc/fasdrc" and "$HOME/.fasdrc"
if they are present.
Below are some variables you can set:
.IP
.nf
\f[C]
$_FASD_DATA
Path\ to\ the\ fasd\ data\ file,\ default\ "$HOME/.fasd".

$_FASD_BLACKLIST
List\ of\ blacklisted\ strings.\ Commands\ matching\ them\ will\ not\ be\ processed.
Default\ is\ "\-\-help".

$_FASD_SHIFT
List\ of\ all\ commands\ that\ needs\ to\ be\ shifted,\ defaults\ to\ "sudo\ busybox".

$_FASD_IGNORE
List\ of\ all\ commands\ that\ will\ be\ ignored,\ defaults\ to\ "fasd\ ls\ echo".

$_FASD_TRACK_PWD
Fasd\ defaults\ to\ track\ your\ "$PWD".\ Set\ this\ to\ 0\ to\ disable\ this\ behavior.

$_FASD_AWK
Which\ awk\ to\ use.\ fasd\ can\ detect\ and\ use\ a\ compatible\ awk.

$_FASD_SINK
File\ to\ log\ all\ STDERR\ to,\ defaults\ to\ "/dev/null".

$_FASD_MAX
Max\ total\ score\ /\ weight,\ defaults\ to\ 2000.

$_FASD_SHELL
Which\ shell\ to\ execute.\ Some\ shells\ will\ run\ faster\ than\ others.\ fasd
runs\ faster\ with\ dash\ and\ ksh\ variants.

$_FASD_BACKENDS
Default\ backends.

$_FASD_RO
If\ set\ to\ any\ non\-empty\ string,\ fasd\ will\ not\ add\ or\ delete\ entries\ from
database.\ You\ can\ set\ and\ export\ this\ variable\ from\ command\ line.

$_FASD_FUZZY
Level\ of\ "fuzziness"\ when\ doing\ fuzzy\ matching.\ More\ precisely,\ the\ number\ of
characters\ that\ can\ be\ skipped\ to\ generate\ a\ match.\ Set\ to\ empty\ or\ 0\ to
disable\ fuzzy\ matching.\ Default\ value\ is\ 2.

$_FASD_VIMINFO
Path\ to\ .viminfo\ file\ for\ viminfo\ backend,\ defaults\ to\ "$HOME/.viminfo"

$_FASD_RECENTLY_USED_XBEL
Path\ to\ XDG\ recently\-used.xbel\ file\ for\ recently\-used\ backend,\ defaults\ to
"$HOME/.local/share/recently\-used.xbel"

$_FASD_NOCASE
Force\ case\ insensitivity\ for\ default\ matching\ with\ value\ 1,\ defaults\ to \0.
\f[]
.fi
.SH DEBUGGING
.PP
Fasd is hosted on GitHub: https://github.com/clvv/fasd
.PP
If fasd does not work as expected, please file a bug report on GitHub
describing the unexpected behavior along with your OS version, shell
version, awk version, sed version, and a log file.
.PP
You can set \f[C]_FASD_SINK\f[] in your \f[C]\&.fasdrc\f[] to obtain a
log.
.IP
.nf
\f[C]
_FASD_SINK="$HOME/.fasd.log"
\f[]
.fi
.SH COPYING
.PP
Fasd is originally written based on code from z
(https://github.com/rupa/z) by rupa deadwyler under the WTFPL license.
Most if not all of the code has been rewritten.
Fasd is licensed under the "MIT/X11" license.
.SH AUTHORS
Wei Dai <x@wei23.net>.


fasd.sh
#!/usr/bin/env sh
# This tool gives you quick access to your frequent/recent files
#
# INSTALL:
#   Source this file somewhere in your shell rc (.bashrc or .zshrc).
#
# SYNOPSIS:
#   fasd [options] [query ...]
#     options:
#       -s        show list of files with their ranks
#       -l        list paths only
#       -i        interactive mode
#       -e <cmd>  set command to execute on the result file
#       -a        match files and directories
#       -d        match directories only
#       -f        match files only
#       -r        match by rank only
#       -h        show a brief help message
#
# EXAMPLES:
#   f foo # list recent files mathcing foo
#   f foo bar # list recent files mathcing foo and bar
#   f -e vim foo # run vim on the most frecent file matching foo
#
# TIPS:
#   alias z="f -d -e cd"
#   alias v="f -e vim"
#   alias m="f -e mplayer"
#   alias o="f -e xdg-open"

fasd() {

  case "$1" in
  --init)
    [ -s "$HOME/.fasdrc" ] && . "$HOME/.fasdrc"
    # set default options
    [ -z "$_F_DATA" ] && _F_DATA="$HOME/.f"
    [ -z "$_F_BLACKLIST" ] && _F_BLACKLIST="--help"
    [ -z "$_F_SHIFT" ] && _F_SHIFT="sudo busybox"
    [ -z "$_F_IGNORE" ] && _F_IGNORE="fasd cd ls echo"
    [ -z "$_F_SINK" ] && _F_SINK=/dev/null
    [ -z "$_F_TRACK_PWD" ] && _F_TRACK_PWD=1
    [ -z "$_F_MAX" ] && _F_MAX=2000

    { if [ -z "$_F_AWK" ]; then
        # awk preferences
        for awk in gawk original-awk nawk mawk awk; do
          $awk "" && _F_AWK=$awk && break
        done
      fi
    } >> "$_F_SINK" 2>&1
    ;;

  --init-alias)
    cat <<EOS
    alias a='fasd -a'
    alias s='fasd -si'
    alias sd='fasd -sid'
    alias sf='fasd -sif'
    alias d='fasd -d'
    alias f='fasd -f'
    z() { [ \$# -gt 0 ] && cd "\$(d -e echo "\$@")" || d "\$@"; }
EOS
  ;;

  --init-interactive)
    { # set default aliases
      eval "$(fasd --init-alias)"
      if compctl; then # zsh
        eval "$(fasd --init-zsh)"
      elif complete; then # bash
        eval "$(fasd --init-bash)"
      else # posix shell
        eval "$(fasd --init-posix)"
      fi
    } >> "$_F_SINK" 2>&1
    ;;

  --init-zsh)
    cat <<EOS
    # zsh command mode completion
    _f_zsh_cmd_complete() {
      local compl
      read -c compl
      compstate[insert]=menu # no expand
      eval 'reply=(\${(f)"\$(fasd --complete "\$compl")"})'
    }
    # enbale command mode completion
    compctl -U -K _f_zsh_cmd_complete -V f -x 'C[-1,-*e],s[-]n[1,e]' -c -- fasd
    # zsh word mode completion
    _f_zsh_word_complete() {
      [ "\$2" ] && local _f_cur="\$2"
      [ -z "\$_f_cur" ] && eval 'local _f_cur="\${words[CURRENT]}"'
      eval 'local fnd="\${_f_cur//,/ }"'
      local typ=\${1:-e}
      fasd --query \$typ \$fnd | sort -nr | sed 's/^[0-9.]*[ ]*//' | while read line; do
        compadd -U -V f "\$line"
      done
      compstate[insert]=menu # no expand
    }
    _f_zsh_word_complete_f() { _f_zsh_word_complete f ; }
    _f_zsh_word_complete_d() { _f_zsh_word_complete d ; }
    _f_zsh_word_complete_trigger() {
      eval 'local _f_cur="\${words[CURRENT]}"'
      # fasd --word-complete-trigger \$_f_cur 1>&2
      eval \$(fasd --word-complete-trigger _f_zsh_word_complete \$_f_cur)
    }
    # enable word mode completion
    zstyle ':completion:*' completer _complete _ignored \
      _f_zsh_word_complete_trigger
    # define zle widgets
    zle -C f-complete 'menu-select' _f_zsh_word_complete
    zle -C f-complete-f 'menu-select' _f_zsh_word_complete_f
    zle -C f-complete-d 'menu-select' _f_zsh_word_complete_d
    # add zsh hook
    _f_preexec() { { eval "fasd --add \$(fasd --sanitize \$3)"; } >> "\$_F_SINK" 2>&1; }
    autoload -U add-zsh-hook
    add-zsh-hook preexec _f_preexec
EOS
    ;;

  --init-bash)
    cat <<EOS
    # bash command mode completion
    _f_bash_cmd_complete() {
      # complete command after "-e"
      eval 'local cur=\${COMP_WORDS[COMP_CWORD]}
      [[ \${COMP_WORDS[COMP_CWORD-1]} == -*e ]] && \
        COMPREPLY=( \$(compgen -A command \$cur) ) && return'
      # get completion results using expanded aliases
      local RESULT=\$( fasd --complete "\$(alias -p \${COMP_WORDS} | \
        sed -n "\\\$s/^.*'\(.*\)'/\1/p") \${COMP_LINE#* }" )
      local IFS=\$'\n'
      eval 'COMPREPLY=( \$RESULT )'
    }
    _f_bash_hook_cmd_complete() {
      for cmd in \$*; do
        complete -F _f_bash_cmd_complete \$cmd
      done
    }
    # enable bash command mode completion
    _f_bash_hook_cmd_complete a s d f sd sf
    # bash word mode completion
    _f_bash_word_complete() {
      [ "\$_f_cur" ] || eval 'local _f_cur="\${COMP_WORDS[COMP_CWORD]}"'
      local typ=\${1:-e}
      eval 'local fnd="\${_f_cur//,/ }"'
      local RESULT=\$(fasd --query \$typ \$fnd | sed 's/^[0-9.]*[ ]*//')
      local IFS=\$'\n'
      eval 'COMPREPLY=( \$RESULT )'
    }
    _f_bash_word_complete_trigger() {
      [ "\$_f_cur" ] || eval 'local _f_cur="\${COMP_WORDS[COMP_CWORD]}"'
      eval "\$(fasd --word-complete-trigger _f_bash_word_complete \$_f_cur)"
    }
    _f_bash_word_complete_wrap() {
      eval 'local _f_cur="\${COMP_WORDS[COMP_CWORD]}"'
      _f_bash_word_complete_trigger
      eval 'local z=\${COMP_WORDS[0]}'
      # try original comp func
      [ "\$COMPREPLY" ] || eval "\$( echo "\$_F_BASH_COMPLETE_P" | \
        sed -n "/ \$z\$/"'s/.*-F \(.*\) .*/\1/p' )"
      # fall back on original complete options
      local cmd="\$(echo "\$_F_BASH_COMPLETE_P" | \
        sed -n "/ \$z\$/"'s/complete/compgen/') \$_f_cur"
      [ "\$COMPREPLY" ] || eval 'COMPREPLY=( \$(eval \$cmd) )'
    }
    _f_bash_hook_word_complete_wrap_all() {
      export _F_BASH_COMPLETE_P="\$(complete -p)"
      for cmd in \$(complete -p | awk '{print \$NF}' | tr '\n' ' '); do
        complete -o default -o bashdefault -F _f_bash_word_complete_wrap \$cmd
      done
    }
    # enable word mode completion as default completion
    complete -o default -o bashdefault -D -F _f_bash_word_complete_trigger
    # add bash hook
    echo \$PROMPT_COMMAND | grep -v -q "fasd --add" && \
      PROMPT_COMMAND='eval "fasd --add \$(fasd --sanitize \$(history 1 | \
      sed -e "s/^[ ]*[0-9]*[ ]*//"))" >> "\$_F_SINK" 2>&1;'"\$PROMPT_COMMAND"
EOS
    ;;

  --init-posix)
    cat <<EOS
    _f_ps1_func() {
      eval "fasd --add \$(fasd --sanitize \$(fc -nl -0 | sed -n '\$s/\s*\(.*\)/\1/p'))"
    }
    _f_ps1_install() {
      echo "\$PS1" | grep -v -q "_f_ps1_func" && \
      export PS1="\\\$(_f_ps1_func >> "\$_F_SINK" 2>&1)\$PS1"
    }
    echo "\$PS1" | grep -q '\\\\' && _f_ps1_install
    [ "\$KSH_VERSION" ] && _f_ps1_install # ksh has the compatibility
EOS
    ;;

  --readlink)
    shift; local p np
    case "$1" in
      /*) p="$1";;
      *) p="$PWD/$1";;
    esac
    np="$(echo "$p" | sed 's@[^/]*/*\.\.\(/\|$\)@@g;s@\./@@g;s@/\+@/@g;s@[./]*$@@')"
    [ -e "${np:=/}" ] || return 1
    echo "$np"
    ;;

  # if "$_f_cur" is a query, then eval all the arguments
  --word-complete-trigger)
    shift
    [ "$2" ] && local _f_cur="$2" || return
    case "$_f_cur" in
      ,*)
        echo "$1" e "$_f_cur";;
      f,*)
        echo "$1" f "${_f_cur#?}";;
      d,*)
        echo "$1" d "${_f_cur#?}";;
      *,,)
        echo "$1" e "$_f_cur";;
      *,,f)
        echo "$1" f "${_f_cur%?}";;
      *,,d)
        echo "$1" d "${_f_cur%?}";;
    esac
    ;;

  --sanitize)
    shift
    echo "$@" | sed 's/\(^\| \).\?[<>|]\+/ /g;s/&$//'
    ;;

  --add) # add entries
    shift

    # stop if we don't own ~/.f (we're another user but our ENV is still set)
    [ -f "$_F_DATA" -a ! -O "$_F_DATA" ] && return

    # make zsh do word splitting here
    [ "$ZSH_VERSION" ] && emulate sh && setopt localoptions

    # blacklists
    local each
    for each in $_F_BLACKLIST; do
      case " $* " in *\ $each\ *) return;; esac
    done

    # shifts
    while true; do
      case " $_F_SHIFT " in
        *\ $1\ *) shift;;
        *) break
      esac
    done

    # ignores
    case " $_F_IGNORE " in
      *\ $1\ *) return
    esac

    shift # shift out the command itself

    local paths
    while [ "$1" ]; do
      # add the adsolute path to "paths", and a separator "|"
      paths="$paths|$(fasd --readlink "$1" 2>> "$_F_SINK")"
      shift
    done

    # add current pwd if the option is set
    [ "$_F_TRACK_PWD" = "1" -a "$PWD" != "$HOME" ] && paths="$paths|$PWD"

    [ -z "${paths##|}" ] && return # stop if we have nothing to add

    # maintain the file
    local tempfile
    tempfile="$(mktemp $_F_DATA.XXXXXX)" || return
    $_F_AWK -v list="$paths" -v now="$(date +%s)" -v max="$_F_MAX" -F"|" '
      BEGIN {
        split(list, files, "|")
        for(i in files) {
          path = files[i]
          if ( path == "" ) continue
          paths[path] = path # array for checking
          rank[path] = 1
          time[path] = now
        }
      }
      $2 >= 1 {
        if( $1 in paths ) {
          rank[$1] = $2 + 1
          time[$1] = now
        } else {
          rank[$1] = $2
          time[$1] = $3
        }
        count += $2
      }
      END {
        if( count > max )
          for( i in rank ) print i "|" 0.9*rank[i] "|" time[i] # aging
        else
          for( i in rank ) print i "|" rank[i] "|" time[i]
      }' "$_F_DATA" 2>> "$_F_SINK" >| "$tempfile"
    if [ $? -ne 0 -a -f "$_F_DATA" ]; then
      env rm -f "$tempfile"
    else
      env mv -f "$tempfile" "$_F_DATA"
    fi
    ;;

  --query)
    shift
    [ "$1" ] && local typ="$1"
    [ "$2" ] && local fnd="$2"
    [ "$3" ] && local mode="$3"
    # query the database, this need some local variables to be set
    while read line; do
      [ -${typ:-e} "${line%%\|*}" ] && echo "$line"
    done < "$_F_DATA" | \
    $_F_AWK -v t="$(date +%s)" -v mode="$mode" -v q="$fnd" -F"|" '
      function frecent(rank, time) {
        dx = t-time
        if( dx < 3600 ) return rank*4
        if( dx < 86400 ) return rank*2
        if( dx < 604800 ) return rank/2
        return rank/4
      }
      function likelihood(pattern, path) {
        m = gsub( "/+", "/", path )
        r = 1
        for( i in pattern ) {
          tmp = path
          gsub( ".*" pattern[i], "", tmp)
          n = gsub( "/+", "/", tmp )
          if( n == m )
            return 0
          else if( n == 0 )
            r *= 20 # F
          else
            r *= 1 - ( n / m )
        }
        return r
      }
      function getRank() {
        if( mode == "rank" )
          f = $2
        else
          f = frecent($2, $3)
        wcase[$1] = f * likelihood( pattern, $1 )
        nocase[$1] = f * likelihood( pattern2, tolower($1) )
      }
      BEGIN {
        split(q, pattern, " ")
        for( i in pattern ) pattern2[i] = tolower(pattern[i]) # nocase
      }
      {
        getRank()
        cx = cx || wcase[$1]
        ncx = ncx || nocase[$1]
      }
      END {
        if( cx ) {
          for( i in wcase )
            if( wcase[i] ) printf "%-10s %s\n", wcase[i], i
        } else if( ncx ) {
          for( i in nocase )
            if( nocase[i] ) printf "%-10s %s\n", nocase[i], i
        }
      }' - 2>> "$_F_SINK"
    ;;

  *) # parsing logic and processing
    [ -f "$_F_DATA" ] || return # no db yet
    local fnd last
    while [ "$1" ]; do case "$1" in
      --complete) [ "$2" = "--" ] && shift; set -- $(echo $2); local list=1 r=r;;
      --) while [ "$2" ]; do shift; fnd="$fnd$1 "; last="$1"; done;;
      -*) local o="${1#-}"; while [ "$o" ]; do case $o in
          s*) local show=1;;
          l*) local list=1;;
          i*) local interactive=1 show=1;;
          r*) local mode=rank;;
          t*) local mode=recent;;
          e*) o="${o#?}"; if [ "$o" ]; then # there are characters after "-e"
                local exec=$o # anything after "-e"
              else # use the next argument
                local exec=${2:?"Argument needed after -e"}
                shift
              fi; break;;
          a*) local typ=e;;
          d*) local typ=d;;
          f*) local typ=f;;
          h*) echo "fasd [options] [query ...]
  options:
    -s        show list of files with their ranks
    -l        list paths only
    -i        interactive mode
    -e <cmd>  set command to execute on the result file
    -a        match files and directories
    -d        match directories only
    -f        match files only
    -r        match by rank only
    -h        show a brief help message" >&2; return;;
        esac; o="${o#?}"; done;;
      *) fnd="$fnd $1"; last="$1";;
    esac; shift; done

    # if we hit enter on a completion just execute
    case "$last" in
     # completions will always start with /
     /*) [ -z "$show$list" -a -${typ:-e} "$last" -a "$exec" ] \
       && $exec "$last" && return;;
    esac

    local result
    result="$(fasd --query 2>> "$_F_SINK")" # query the database
    [ $? -gt 0 ] && return
    if [ "$interactive" ]; then
      result="$(echo "$result" | sort -nr)"
      echo "$result" | sed = | sed 'N;s/\n/\t/' | sort -nr
      local i; printf "> "; read i; [ 0 -lt "$i" ] || exit 1
      ${exec:=echo} "$(echo "$result" | sed -n "${i:=1}"'s/^[0-9.]*[ ]*//p')"
    elif [ "$list" ]; then
      echo "$result" | sort -n${r} | sed 's/^[0-9.]*[ ]*//'
    elif [ "$show" ]; then
      echo "$result" | sort -n${r}
    elif [ "$fnd" -a "$exec" ]; then # exec
      $exec "$(echo "$result" | sort -n | sed -n '$s/^[0-9.]*[ ]*//p')"
    elif [ "$fnd" -a ! -t 1 ]; then # echo if output is not terminal
      echo "$result" | sort -n | sed -n '$s/^[0-9.]*[ ]*//p'
    else # no args, show
      echo "$result" | sort -n${r}
    fi

  esac
}

fasd --init

case "$-" in
  *i*) fasd --init-interactive;; # assume being sourced
  *) # assume being executed as an executable
    if [ -x "$_F_SHELL" -a -z "$_F_SET" ]; then
      _F_SET=1 $_F_SHELL "$0" "$@"
      exit $?
    else
      fasd "$@"
    fi
esac

